### PR TITLE
Utilize `webots_ros2_driver` in `webots_ros2_epuck`

### PR DESCRIPTION
**Pull request by BenjaminHug8**
**Description**
The objective of this PR is to deprecate the `webots_ros2_core` packages in favor of `webots_ros2_driver` for `webots_ros2_epuck`.

**Tasks**
- [x] Convert
- [x] Add system tests

### CODE REVIEW

File: webots_ros2_epuck/resource/epuck_webots.urdf [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        <joint name="right wheel motor">
+            <state_interface name="position"/>
+            <command_interface name="velocity"/>
+        </joint>
+        <joint name="left wheel motor">
+            <state_interface name="position"/>
+            <command_interface name="velocity"/>
+        </joint>
+    </ros2_control>
+</robot>
```
Comment (lukicdarkoo) [inline webots_ros2_epuck/resource/epuck_webots.urdf:84]:
```suggestion
</robot>

```

File: webots_ros2_epuck/setup.py [hunk 2]
```diff
@@ -64,7 +66,7 @@
     tests_require=['pytest'],
     entry_points={
         'console_scripts': [
-            'driver = webots_ros2_epuck.driver:main',
+            'epuck_process = webots_ros2_epuck.epuck_node:main',
```
Comment (lukicdarkoo) [inline webots_ros2_epuck/setup.py:69]:
```suggestion
            'epuck_node = webots_ros2_epuck.epuck_node:main',
```
I would keep it consistent

File: webots_ros2_epuck/webots_ros2_epuck/epuck_node.py [hunk 3]
```diff
@@ -46,47 +42,33 @@
...
-        self.distance_sensors = {}
+        self.__subscriber_dist_sensors = {}
+        self.__dists = {}
         for i in range(NB_INFRARED_SENSORS):
-            sensor = self.robot.getDistanceSensor('ps{}'.format(i))
-            sensor.enable(self.timestep)
-            self.distance_sensors['ps{}'.format(i)] = sensor
+            self.__dists['ps{}'.format(i)] = OUT_OF_RANGE
+
+        self.__subscriber_dist_sensors['ps0'] = self.create_subscription(Range, '/e_puck/ps0', self.__process_dist_sens_0, 1)
```
Comment (lukicdarkoo) [inline webots_ros2_epuck/webots_ros2_epuck/epuck_node.py:56]:
```suggestion
        for i in range(7):
            self.__subscriber_dist_sensors[f'ps{i}'] = self.create_subscription(Range, f'/e_puck/ps{i}', lambda message: self.__on_distance_sensor_message(message, i), 1)
```
I believe something like this is possible (maybe slightly different)

File: webots_ros2_epuck/webots_ros2_epuck/epuck_node.py [hunk 4]
```diff
@@ -46,47 +42,33 @@
...
-        self.start_device_manager(DEVICE_CONFIG)
+class EPuckNode(Node):
+    def __init__(self):
+        super().__init__('epuck_node')
+        self.get_logger().info("Epuck node has been started.")
 
         # Intialize distance sensors for LaserScan topic
-        self.distance_sensors = {}
+        self.__subscriber_dist_sensors = {}
+        self.__dists = {}
```
Comment (lukicdarkoo) [inline webots_ros2_epuck/webots_ros2_epuck/epuck_node.py:52]:
```suggestion
        self.__distances = {}
```
Avoid abbrevations

File: webots_ros2_epuck/webots_ros2_epuck/epuck_node.py [hunk 5]
```diff
@@ -46,47 +42,33 @@
...
+        self.__subscriber_dist_sensors['ps1'] = self.create_subscription(Range, '/e_puck/ps1', self.__process_dist_sens_1, 1)
+        self.__subscriber_dist_sensors['ps2'] = self.create_subscription(Range, '/e_puck/ps2', self.__process_dist_sens_2, 1)
+        self.__subscriber_dist_sensors['ps3'] = self.create_subscription(Range, '/e_puck/ps3', self.__process_dist_sens_3, 1)
+        self.__subscriber_dist_sensors['ps4'] = self.create_subscription(Range, '/e_puck/ps4', self.__process_dist_sens_4, 1)
+        self.__subscriber_dist_sensors['ps5'] = self.create_subscription(Range, '/e_puck/ps5', self.__process_dist_sens_5, 1)
+        self.__subscriber_dist_sensors['ps6'] = self.create_subscription(Range, '/e_puck/ps6', self.__process_dist_sens_6, 1)
+        self.__subscriber_dist_sensors['ps7'] = self.create_subscription(Range, '/e_puck/ps7', self.__process_dist_sens_7, 1)
+
+        self.__subscriber_tof = self.create_subscription(Range, '/e_puck/tof', self.__process_tof, 1)
+        self.__new_tof_data = False
```
Comment (lukicdarkoo) [inline webots_ros2_epuck/webots_ros2_epuck/epuck_node.py:66]: We can treat it as the other distances sensors (we keep the value in `self.__distances`). If `self.__distances['tof'] == OUT_OF_RANGE` than the value will be automatically ignored, no special "treatment" is needed.

**Review by lukicdarkoo — CHANGES_REQUESTED** [review]
In general, it looks very good!

File: webots_ros2_epuck/webots_ros2_epuck/epuck_node.py (hunk 3 above)
Comment (BenjaminHug8) [reply webots_ros2_epuck/webots_ros2_epuck/epuck_node.py:56]: Thanks, I had no idea if it was possible to configure the name of the callback

File: webots_ros2_epuck/resource/epuck_webots.urdf [hunk 6]
```diff
@@ -0,0 +1,84 @@
+<?xml version="1.0" ?>
+<robot name="Epuck Webots">
+    <webots>
+
+        <device reference="camera" type="Camera">
+            <ros>
+                <enabled>true</enabled>
+                <alwaysOn>true</alwaysOn>
```
[the remark this answers is not in the document]
Comment (BenjaminHug8) [reply webots_ros2_epuck/resource/epuck_webots.urdf:8]: Ok I will try it !

File: webots_ros2_epuck/webots_ros2_epuck/epuck_node.py (hunk 5 above)
Comment (BenjaminHug8) [reply webots_ros2_epuck/webots_ros2_epuck/epuck_node.py:67]: Hum, so you would to put tof in `self.__distances['tof']`, but then it has to not go through the check `laser_dists = [OUT_OF_RANGE if dist > INFRARED_MAX_RANGE else dist for dist in dists]` as ToF has another `INFRARED_MAX_RANGE`, right ?

File: webots_ros2_epuck/resource/epuck_webots.urdf (hunk 6 above)
[the remark this answers is not in the document]
Comment (BenjaminHug8) [reply webots_ros2_epuck/resource/epuck_webots.urdf:8]: It works

File: webots_ros2_epuck/launch/robot_launch.py [hunk 7]
```diff
@@ -17,35 +17,113 @@
...
-            ('executable', 'driver'),
-            ('world', PathJoinSubstitution([package_dir, 'worlds', world])),
+    webots = WebotsLauncher(
+        world=PathJoinSubstitution([package_dir, 'worlds', world])
+    )
+
+    # TODO: Revert once the https://github.com/ros-controls/ros2_control/pull/444 PR gets into the release
+    sleep_time = 60 if 'CI' in os.environ else 10
+    controller_manager_timeout = ['--controller-manager-timeout', '50'] if os.name == 'nt' else []
+    controller_manager_prefix = 'python.exe' if os.name == 'nt' else f"bash -c 'sleep {sleep_time}; $0 $@' "
```
Comment (lukicdarkoo) [inline webots_ros2_epuck/launch/robot_launch.py:45]: https://github.com/ros-controls/ros2_control/pull/444 is released. We can use it as here:
https://github.com/cyberbotics/webots_ros2/blob/2d175978bfd3382dbce58b266532d6a111a36a24/webots_ros2_universal_robot/launch/robot_launch.py#L45-L46

File: webots_ros2_epuck/launch/robot_with_tools_launch.py [hunk 8]
```diff
@@ -27,16 +27,16 @@
...
     package_dir = get_package_share_directory('webots_ros2_epuck')
 
-    use_nav = LaunchConfiguration('nav', default=False)
-    use_rviz = LaunchConfiguration('rviz', default=False)
-    use_mapper = LaunchConfiguration('mapper', default=False)
-    synchronization = LaunchConfiguration('synchronization', default=False)
+    use_nav = LaunchConfiguration('nav', default=True)
+    use_rviz = LaunchConfiguration('rviz', default=True)
+    use_mapper = LaunchConfiguration('mapper', default=True)
+    synchronization = LaunchConfiguration('synchronization', default=True)
```
Comment (lukicdarkoo) [inline webots_ros2_epuck/launch/robot_with_tools_launch.py:33]: Why are these changed?

File: webots_ros2_epuck/resource/epuck_webots.urdf [hunk 9]
```diff
@@ -0,0 +1,83 @@
+<?xml version="1.0" ?>
+<robot name="Epuck Webots">
+    <webots>
+
+        <device reference="camera" type="Camera">
+            <ros>
+                <enabled>true</enabled>
+            </ros>
+        </device>
```
Comment (lukicdarkoo) [inline webots_ros2_epuck/resource/epuck_webots.urdf:9]:
```suggestion
```
It should be enabled by default, we can delete the block

**Review by lukicdarkoo — CHANGES_REQUESTED** [review]

File: webots_ros2_epuck/launch/robot_launch.py (hunk 7 above)
Comment (BenjaminHug8) [reply webots_ros2_epuck/launch/robot_launch.py:45]: Ok, I have not noticed it, thanks!

File: webots_ros2_epuck/launch/robot_with_tools_launch.py (hunk 8 above)
Comment (BenjaminHug8) [reply webots_ros2_epuck/launch/robot_with_tools_launch.py:33]: It was to for testing, forget to revert.

**Review by lukicdarkoo — CHANGES_REQUESTED** [review]
I have just tested and it works well. There is only one thing remaining to be fixed.

These are topics published by e-puck:
```
/e_puck/camera
/e_puck/camera/camera_info
/e_puck/led0
/e_puck/led1
/e_puck/led2
/e_puck/led3
/e_puck/led4
/e_puck/led5
/e_puck/led6
/e_puck/led7
/e_puck/led8
/e_puck/led9
/e_puck/ls0
/e_puck/ls1
/e_puck/ls2
/e_puck/ls3
/e_puck/ls4
/e_puck/ls5
/e_puck/ls6
/e_puck/ls7
/e_puck/pi_puck_led_0
/e_puck/pi_puck_led_1
/e_puck/pi_puck_led_2
/e_puck/ps0
/e_puck/ps1
/e_puck/ps2
/e_puck/ps3
/e_puck/ps4
/e_puck/ps5
/e_puck/ps6
/e_puck/ps7
/e_puck/tof
```
and those differ from ones in the documentation:
https://github.com/cyberbotics/webots_ros2/wiki/Tutorial-E-puck-for-ROS2-Beginners

There is a `<topicName>` tag that allows you to specify a custom topic name:
https://github.com/cyberbotics/webots_ros2/wiki/Tutorial-Creating-a-Custom-Package#urdf-webots-tag

**Comment by BenjaminHug8** [comment]
> I have just tested and it works well. There is only one thing remaining to be fixed.
> 
> These are topics published by e-puck:
> 
> ```
> /e_puck/camera
> /e_puck/camera/camera_info
> /e_puck/led0
> /e_puck/led1
> /e_puck/led2
> /e_puck/led3
> /e_puck/led4
> /e_puck/led5
> /e_puck/led6
> /e_puck/led7
> /e_puck/led8
> /e_puck/led9
> /e_puck/ls0
> /e_puck/ls1
> /e_puck/ls2
> /e_puck/ls3
> /e_puck/ls4
> /e_puck/ls5
> /e_puck/ls6
> /e_puck/ls7
> /e_puck/pi_puck_led_0
> /e_puck/pi_puck_led_1
> /e_puck/pi_puck_led_2
> /e_puck/ps0
> /e_puck/ps1
> /e_puck/ps2
> /e_puck/ps3
> /e_puck/ps4
> /e_puck/ps5
> /e_puck/ps6
> /e_puck/ps7
> /e_puck/tof
> ```
> 
> and those differ from ones in the documentation:
> https://github.com/cyberbotics/webots_ros2/wiki/Tutorial-E-puck-for-ROS2-Beginners
> 
> There is a `<topicName>` tag that allows you to specify a custom topic name:
> https://github.com/cyberbotics/webots_ros2/wiki/Tutorial-Creating-a-Custom-Package#urdf-webots-tag

I was not aware of this, I will make the changes and keep it in mind for future work!

**Review by lukicdarkoo — APPROVED** [review]
Great, thanks!
